athena-dynamodb/athena-dynamodb.yaml

```diff
@@ -45,35 +52,94 @@ Resources:
           disable_spill_encryption: !Ref DisableSpillEncryption
           spill_bucket: !Ref SpillBucket
           spill_prefix: !Ref SpillPrefix
+          kms_key_id: !If [HasKMSKeyId, !Ref KMSKeyId, !Ref "AWS::NoValue"]
       FunctionName: !Ref AthenaCatalogName
       Handler: "com.amazonaws.athena.connectors.dynamodb.DynamoDBCompositeHandler"
       CodeUri: "./target/athena-dynamodb-1.0.jar"
       Description: "Enables Amazon Athena to communicate with DynamoDB, making your tables accessible via SQL"
       Runtime: java8
       Timeout: !Ref LambdaTimeout
       MemorySize: !Ref LambdaMemory
-      Policies:
-        - Statement:
-            - Action:
-                - dynamodb:DescribeTable
-                - dynamodb:ListSchemas
-                - dynamodb:ListTables
-                - dynamodb:Query
-                - dynamodb:Scan
-                - glue:GetTableVersions
-                - glue:GetPartitions
-                - glue:GetTables
-                - glue:GetTableVersion
-                - glue:GetDatabases
-                - glue:GetTable
-                - glue:GetPartition
-                - glue:GetDatabase
-                - athena:GetQueryExecution
-                - s3:ListAllMyBuckets
-              Effect: Allow
-              Resource: '*'
-          Version: '2012-10-17'
-        #S3CrudPolicy allows our connector to spill large responses to S3. You can optionally replace this pre-made policy
-        #with one that is more restrictive and can only 'put' but not read,delete, or overwrite files.
-        - S3CrudPolicy:
-            BucketName: !Ref SpillBucket
\ No newline at end of file
+      Role: !GetAtt FunctionRole.Arn
+
+  FunctionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      ManagedPolicyArns:
+        - "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
+      AssumeRolePolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service:
+                - lambda.amazonaws.com
+            Action:
+              - "sts:AssumeRole"
```

Comment:
Why did we change this?

Reply:
I had to explicitly create a role for the lambda instead of using the existing Policies property in the Lambda because using Policies there is no way to define the Resource part of:

          - Action:
              - kms:GenerateDataKey
            Effect: Allow
            Resource: !Sub "arn:aws:kms:${AWS::Region}:${AWS::AccountId}:key/${KMSKeyId}"

for both the cases where KMSKeyId is passed and when it is not, because when KMSKeyId is not passed the above statement fails and to avoid that instead, when KMSKeyId is not passed I just do not create the FunctionKMSPolicy at all using the condition "HasKMSKeyId". So to recap I had to modify the role\policy creation from the previous way that implicitly created them via just listing the permissions under the lambda Policies key to explicitly create them so that I can avoid to create the FunctionKMSPolicy if KMSKeyId is not passed and I would not be able to define the Resource filter for the kms:GenerateDataKey permission. Hope it makes sense but happy to explain it further if required.

Reply:
Ack.